src/test/java/org/elasticsearch/cluster/metadata/MetaDataTests.java

```diff
@@ -579,76 +592,145 @@ public void concreteIndicesWildcardExpansion() {
         assertThat(newHashSet(md.concreteIndices(IndicesOptions.fromOptions(true, true, true, true), "testX*")), equalTo(newHashSet("testXXX", "testXXY", "testXYY")));
     }
 
+    /**
+     * test resolving _all pattern (null, empty arry or "_all") for random IndicesOptions
+     */
+    @Test
+    public void concreteIndicesAllpatternRandom() {
+        for (int i = 0; i < 100; i++) {
+            String[] allIndices = null;
+            switch (randomIntBetween(0, 2)) {
+            case 0:
+                break;
+            case 1:
+                allIndices = new String[0];
+                break;
+            case 2:
+                allIndices = new String[] { MetaData.ALL };
+                break;
+            }
+
+            IndicesOptions indicesOptions = IndicesOptions.fromOptions(randomBoolean(), randomBoolean(), randomBoolean(), randomBoolean());
+            MetaData metadata = MetaData.builder().build();
+
+            // with no indices, asking for all indices should return empty list or exception, depending on indices options
+            try {
+                String[] concreteIndices = metadata.concreteIndices(indicesOptions, allIndices);
+                assertThat(concreteIndices, notNullValue());
+                assertThat(concreteIndices.length, equalTo(0));
+            } catch (IndexMissingException e) {
+                if (indicesOptions.allowNoIndices())
+                    fail("calling concreteIndices for all indices with allowNoIndices option should not throw exception");
+            }
```

Comment:
I think it is not clear here exactly when IndexMissingException is expected to be thrown or not. I would rather move the if on top and have different asserts path based on that. FOr the expected exception one you can then do:

```
try {
  //do something
  fail("shouldn't get here");
} catch (IndexMissingException e) {
  //assert on exception
}

```